FIX FOR ISSUE #3612

Anjali-NEC opened the pull request:
Upgraded built-from-sources dependencies to newer version fixed for issue #3612

fgalan commented on ci/rpm7/build-dep.sh:
```diff
@@ -51,16 +51,16 @@ curl -L https://github.com/mongodb/mongo-cxx-driver/archive/legacy-1.1.2.tar.gz
-curl -L http://ftp.gnu.org/gnu/libmicrohttpd/libmicrohttpd-0.9.48.tar.gz | tar xzC /opt/ \
-&& cd /opt/libmicrohttpd-0.9.48  \
+curl -L http://ftp.gnu.org/gnu/libmicrohttpd/libmicrohttpd-0.9.69.tar.gz | tar xzC /opt/ \
+&& cd /opt/libmicrohttpd-0.9.69  \
```
There is a newer version (0.9.70). Maybe that one could be used.

fgalan commented:
Thanks for the conbritution!

I have one general doubt...

Did you test a from-scrach build of Orion based in the new procedure? Or you only changed the .md numbers in the files?

fgalan commented:
> Hi @fgalan sir,
> Yes, I have tested it from-scrach build of Orion based in the new procedure.

Ok. I'll use your procedure to "upgrade" my CentOS 7 reference image (the place from where I produce the Orion official RPMs) and, at the same time, provide an extra validation.

Thanks!

fgalan commented on ci/rpm7/build-dep.sh:
```diff
@@ -51,16 +51,16 @@ curl -L https://github.com/mongodb/mongo-cxx-driver/archive/legacy-1.1.2.tar.gz
 && scons install --disable-warnings-as-errors --prefix=/usr/local --use-sasl-client --ssl \
```
The upgrade done by this PR should be described in CHANGE_NEXT_RELEASE. For instance:

```
- Fix: upgrade rapidjson dependecy from 1.0.2 to 1.1.0
- Fix: upgrade microhttpd dependency from 0.9.48 to 0.9.70
```

fgalan commented:
> Ok. I'll use your procedure to "upgrade" my CentOS 7 reference image (the place from where I produce the Orion official RPMs) and, at the same time, provide an extra validation.

I have done this process. Although I have been able to build Orion successfully and pass unit tests (`make unit_test`) I have found problems with functional tests (`make ft`).

It seems that some small things have changed between microhttpd 0.9.48 to 0.9.70. I have found two differences:

* Previously responses without payload include `Content-Length: 0` header. Now, this header is omitted.
* Previously `HTTP/1.1 413 Request Entity Too Large` was used. Now, `HTTP/1.1 413 Payload Too Large` is used.

@Anjali-NEC could you run functional tests in your environment to check if the result is the same in your environment?

If confirmed, two things need to be done:

* First, confirm these changes are not a problem from the point of view of backward compatibility of client applications using Orion. Maybe analyzing the cause of the changes in the microhttpd changelog from 0.9.48 to 0.9.70 could provide some insight.
* In the case there isn't any problem from the point of view of backward compatibility, .test files need to be changed in this PR to align with the way microhttpd works now.

fgalan commented:
(Note in this PR we cannot rely on the test done by travis. Travis test is based on the current dockerhub CI image, which hasn't incorporated yet the upgrade library changes. This is the reason because travis passes but if you run the same in your local environment it shouldn't pass).

Anjali-NEC commented:
Hi @fgalan sir,

I have run the functional tests in my local environment. There are many test cases are failing due to changes in versions of microhttpd and rapidjson.

> First, confirm these changes are not a problem from the point of view of backward compatibility of client applications using Orion. Maybe analyzing the cause of the changes in the microhttpd changelog from 0.9.48 to 0.9.70 could provide some insight.

> In the case there isn't any problem from the point of view of backward compatibility, .test files need to be changed in this PR to align with the way microhttpd works now.

As per my understanding, to support updated versions, we need to modify the test case.

fgalan commented:
> As per my understanding, to support updated versions, we need to modify the test case.

Before modifying anything it would be great to understand the cause of the changes and if they could be some problem from the point of view of backward compatibility in existing client software using Orion. Could you have a look to the microhttpd changelog from 0.9.48 to 0.9.70 and identify the entries corresponding to the change we have detected (ChangeLog file inside the .tar.gz)?

In addition, have you detected changes related with the rapidjson upgrade (as far as I see all the changed were due to microhttpd upgrade, but I could have missed something)?

Anjali-NEC commented:
Hi @fgalan sir,

I've investigated the root cause of error occurring in test cases. Following are the details of cause from the changelog:
1. In mircohttpd version 0.9.49 ChnageLog  -
"Fixed handling of misformed URI with spaces".

2. In mircohttpd version 0.9.51 ChnageLog -
"Do not send "Content-Length" header for 1xx/204/304 status codes. -CG"

3. In mircohttpd version 0.9.55 ChnageLog -
"MHD_HTTP_REQUEST_HEADER_FIELDS_TOO_LARGE (431) instead of old
 MHD_HTTP_REQUEST_ENTITY_TOO_LARGE (413) for very long header".

I have fixed the errors in all test cases which occurred due to above changelog. I have also added all modification in current PR.
Now, all test cases has been passed on my local environment except one test case 3658_env_vars/env_vars.test. As per my analysis, this test case is also getting failed in previous version i.e. 0.9.48


>In addition, have you detected changes related with the rapidjson upgrade (as far as I see all the changed were due to microhttpd upgrade, but I could have missed something)?

No, there is no change related to the rapidjson upgrade

fgalan commented on test/functionalTest/cases/0000_large_requests/large_discover_context_availability.test:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2013 Telefonica Investigacion y Desarrollo, S.A.U
+#	 Copyright 2013 Telefonica Investigacion y Desarrollo, S.A.U
```
```suggestion
# Copyright 2013 Telefonica Investigacion y Desarrollo, S.A.U
```

fgalan replied:
Fixed in eea62e9

fgalan commented:
@Anjali-NEC thanks you for your investigation work!

One doubt, with regards to:

> "Fixed handling of misformed URI with spaces".

I'm not sure to which test it affects. Could you tell the .test files in which this change is impacting, please? (Or, in other words, the .test that needed to be changed due to that change in microhttpd).

Anjali-NEC commented:
Hi @fgalan sir,

It affects only one test case i.e ```1705_csub_cache_objects/csub_cache_objects_errors.test```. I have updated this test case as per the latest version of microhttpd.

fgalan commented:
I think the modifications in the .test are fine. Note that we cannot rely yet on the test check done by jenkins, as the docker image in which it is based has not been yet updated with the new versions of the library (it will be done after this the LGTM on this PR).

Thus, by the moment, let's rely on local test pass. I did one on commit b58a93f8fa6b6c05dd6cf769b0262f4e81130fc9 and I find the following test yet failing:

```
----------- Failing tests ------------------
  o  0000_deprecated_checkings/uppercase_action_types_in_ngsiv2_batch_update.test
  o  0740_crash_with_json_object_for_attributes/json_object_format.test
  o  0876_entity_dates/entity_dates_overridden_by_user.test
  o  0876_entity_dates/entity_dates_overridden_by_user_subs.test
  o  0876_entity_dates/filter_datemodified_sub.test
  o  0876_entity_dates/filter_entity_dates.test
  o  0876_entity_dates/get_entity_dates.test
  o  0876_entity_dates/notify_entity_dates_with_attr.test
  o  1038_ngsiv2_geo_location_entities/errors_general.test
  o  1490_compound_values_with_dot_in_key/compound_values_with_dot_in_key.test
```

I think the following ones needs "zero content-lenght removal fix":

*  0000_deprecated_checkings/uppercase_action_types_in_ngsiv2_batch_update.test (maybe already fixed in later commit eea62e9 ?)
*  0876_entity_dates/entity_dates_overridden_by_user.test
*  0876_entity_dates/entity_dates_overridden_by_user_subs.test
*  0876_entity_dates/filter_datemodified_sub.test
*  0876_entity_dates/filter_entity_dates.test
*  0876_entity_dates/get_entity_dates.test
*  0876_entity_dates/notify_entity_dates_with_attr.test
*  1490_compound_values_with_dot_in_key/compound_values_with_dot_in_key.test

I think the following ones needs "Payload Too Large fix"

* 1038_ngsiv2_geo_location_entities/errors_general.test

Which regards to 0740_crash_with_json_object_for_attributes/json_object_format.test I'm not sure what the problem is...

fgalan commented:
I have been analyzing potential problems due to the changes introduced by libmicrohttpd.

---------------------------

Regarding the removal of "Content-Lenght: 0" in some responses, from [HTTP Spec RFC7230 section 3.3.2](https://tools.ietf.org/html/rfc7230#section-3.3.2):

> A server MUST NOT send a Content-Length header field in any response with a status code of 1xx (Informational) or 204 (No Content)

It can be checked that all the modifications to remove "Content-Length: 0" in .test are in 204 responses. Thus, this was a bug in previous versions of libmicrohttpd version.

-----------------------------

Regarding the change of the status code for 413 responses from "Request Entity Too Large" to "Payload Too Large", from [HTTP Spec RFC7230 section 6.1](https://tools.ietf.org/html/rfc7231#section-6.1)

>    | 413  | Payload Too Large             |

That section of the spec, in addition says:

> The reason phrases listed here are only recommendations

Thus, no client should rely on specific wording of reason phrases (they are aimed at humans), clients should use the response code itself (413), which hasn't changed.

------------------------------

Regarding the changes in 1705_csub_cache_objects/csub_cache_objects_errors.test they are improving error detection, so I don't see problems with them.

-----------------------------

**Conclusions**: no issue related with backward compatibility due to the upgrade of libmicrohttpd version (on the contrary, the upgrade will make Orion behaver better, from the point of view of the HTTP protocol).

@Anjali-NEC what do you think? Do you agree with my analysis? Or do you see any problem?

fgalan commented:
@fisuda given this PR is modifying the Japanese translation document, it will be great if you could provide feedback on it (LGTM if the modifications are ok or feedback if they aren't). Thanks!

fisuda commented:
@fgalan I have just reviewed the Japanese document. LGTM.
@Anjali-NEC Thank you for updating the Japanese document.

Anjali-NEC commented:
> Thus, by the moment, let's rely on local test pass. I did one on commit [b58a93f](https://github.com/telefonicaid/fiware-orion/commit/b58a93f8fa6b6c05dd6cf769b0262f4e81130fc9) and I find the following test yet failing:
> 
> ```
> ----------- Failing tests ------------------
>   o  0000_deprecated_checkings/uppercase_action_types_in_ngsiv2_batch_update.test
>   o  0740_crash_with_json_object_for_attributes/json_object_format.test
>   o  0876_entity_dates/entity_dates_overridden_by_user.test
>   o  0876_entity_dates/entity_dates_overridden_by_user_subs.test
>   o  0876_entity_dates/filter_datemodified_sub.test
>   o  0876_entity_dates/filter_entity_dates.test
>   o  0876_entity_dates/get_entity_dates.test
>   o  0876_entity_dates/notify_entity_dates_with_attr.test
>   o  1038_ngsiv2_geo_location_entities/errors_general.test
>   o  1490_compound_values_with_dot_in_key/compound_values_with_dot_in_key.test
> ```


I have updated these test cases in my latest commit, they are passing now. 

> Which regards to 0740_crash_with_json_object_for_attributes/json_object_format.test I'm not sure what the problem is...

This test case is passed on my local environment.

Anjali-NEC commented:
> @Anjali-NEC what do you think? Do you agree with my analysis? Or do you see any problem?

Yes, I agree with your analysis, I don't see any problem in it.

fgalan commented:
Travis got green! Merging...

Thanks!
